File: src/print-in-order/print-in-order.py
from threading import Event

class Foo:
    def __init__(self):
        pass
        self.firstDone = Event()
        self.secondDone = Event()


    def first(self, printFirst: 'Callable[[], None]') -> None:
        
        # printFirst() outputs "first". Do not change or remove this line.
        printFirst()
        self.firstDone.set()


    def second(self, printSecond: 'Callable[[], None]') -> None:
        self.firstDone.wait()
        # printSecond() outputs "second". Do not change or remove this line.
        printSecond()
        self.secondDone.set()


    def third(self, printThird: 'Callable[[], None]') -> None:
        self.secondDone.wait()
        # printThird() outputs "third". Do not change or remove this line.
        printThird()

File: src/print-in-order/print-in-order.java
import java.util.concurrent.CountDownLatch;

class Foo {
    private final CountDownLatch latchFirst = new CountDownLatch(1);
    private final CountDownLatch latchSecond = new CountDownLatch(1);

    public Foo() {
        
    }

    public void first(Runnable printFirst) throws InterruptedException {
        
        // printFirst.run() outputs "first". Do not change or remove this line.
        printFirst.run();
        latchFirst.countDown();
    }

    public void second(Runnable printSecond) throws InterruptedException {
        latchFirst.await();
        // printSecond.run() outputs "second". Do not change or remove this line.
        printSecond.run();
        latchSecond.countDown();
    }

    public void third(Runnable printThird) throws InterruptedException {
        latchSecond.await();
        // printThird.run() outputs "third". Do not change or remove this line.
        printThird.run();
    }
}